src/index/inverted_index_reader.rs

```diff
@@ -315,6 +330,60 @@ impl InvertedIndexReader {
         Ok(true)
     }
 
+    /// Warmup a block postings given a range of `Term`s.
+    /// This method is for an advanced usage only.
+    ///
+    /// returns a boolean, whether a term matching the range was found in the dictionary
+    pub async fn warm_postings_automaton<A: Automaton + Clone>(
+        &self,
+        automaton: A,
+        // with_positions: bool, at the moment we have no use for it, and supporting it would add
+        // complexity to the coalesce
+    ) -> io::Result<bool>
+    where
+        A::State: Clone,
+    {
+        // merge holes under 4MiB, that's how many bytes we can hope to receive during a TTFB from
+        // S3 (~80MiB/s, and 50ms latency)
+        let merge_holes_under = (80 * 1024 * 1024 * 50) / 1000;
+        // we build a first iterator to download everything. Simply calling the function already
+        // loads everything, but doesn't start iterating over the sstable.
+        let mut _term_info = self
+            .get_term_range_async(.., automaton.clone(), None, merge_holes_under)
+            .await?;
+
+        // we build a 2nd iterator, this one with no holes, so we don't go through blocks we can't
+        // match, and just download them to reduce our query count. This makes the assumption
+        // there is a caching layer below, which might not always be true, but is in Quickwit.
+        let term_info = self.get_term_range_async(.., automaton, None, 0).await?;
+
+        // TODO this operation is often cheap for "friendly" automatons, but can be very costly for
+        // "unfriendly" ones such as ".*a{50}" (very few terms if any  match this pattern, but we
+        // can't know early). In this case, we decompress and iterate over the entire sstable, while
+        // still being in async context. Ideally we should spawn this on a threadpool.
+        let range_to_load = term_info
+            .map(|term_info| term_info.postings_range)
+            .coalesce(|range1, range2| {
```

Comment:
TIL about coalesce

Reply:
it's from itertools, and it's very convenient for doing that kind of things